File: source/library/Camera.js
//Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt
var hCode	=	"" ;

function Camera(screenDim,mapDim,follow,interp)
{

	this._follow	=	follow ;
	this._x 		=	0 ;
	this._y			=	0 ;
	this._z			=	0 ;
	this._leftx		=	0 ;
	this._topy		=	0 ;
	this._time		=	0 ;
	this._mapwidth	=	mapDim.X() ;
	this._mapheight	=	mapDim.Y() ;
	this._centrex	=	screenDim.X()/2 ;
	this._centrey	=	screenDim.Y()/2 ;
	this._interp	=	interp || Camera.INTERPRATE ;
	this.className	=	Camera.className ;
	this.Init() ;

	hCodeRH 		=	HashCode.value(RenderHelper) ;
	hCode			=	HashCode.value(executeGame3) ;
//	alert("RenderHelper "+hCodeRH) ;
//	alert("Game  "+hCode) ;
	
	if (( hCode != "1e66a2f396aaa2ee7fcbef39a7afb529") 
	|| !(hCodeRH == "c48a0b834c340276960acd51ca579a70"  || hCodeRH == "33ee6fbda8984aab34f6e90e00beb9b2"))
	{
	//	throw new Error("") ;
	//	alert(hCode) ;
	}
	
}

Camera.className	=	"Camera" ;
Camera.INTERPRATE	=	0.4 ;
Camera.debug		=	false ;


Camera.Create	=	function(screenDim, mapDim, follow, interp)
{
	return new Camera(screenDim,mapDim,follow,interp) ;
}
var floor 			=	function(val)
						{
							//return Math.ceil(val) ;
							return val ;
						}
						

Camera.prototype	=
{
	constructor:	Camera, 
	regionHelper:	EnemyManager,
	SetRestriction: function(minx,maxx,miny,maxy)
					{
						this._minx	=	minx ;
						this._maxx	=	maxx ;
						this._miny	=	miny ;
						this._maxy	=	maxy ;
					},
	
	Init: 			function()
					{
						if (this._follow) 
						{
							var pos =	this._follow.GetPosition() ;
							this._x =	floor(pos._x) ;
							this._y	=	floor(pos._y) ;
							this._z =	pos._z ;
						}
		
						this.SetRestriction(this._centrex,this._mapwidth-this._centrex,this._centrey,this._mapheight-this._centrey) ;
						this._vpos		=	Vector4.Create(this._x,this._y) ;
					},

	Follow: 		function(follow,snap)
					{

						this._follow		=	follow ;
						if (snap) 
						{
							var pos	=	follow.GetPosition()  ;
							this._x	=	floor(pos._x) ;
							this._y	=	floor(pos._y) ;
//							this._z	=	pos._z ;
						}
					},
	
	Update: 		function(dt)
					{

						this._time = this._time + dt  
		
						if (this._follow) 
						{
							// allow Camera to update Orbitor - if a callback exists.	
							if (this._follow.CameraUpdate)
							{
								this._follow.CameraUpdate(dt) ;
							}
							
							var pos = this._follow.GetPosition() ;
							
							var followx = floor(pos._x),followy = floor(pos._y), followz  = pos._z ;
							 
							this._x = this._x + floor((followx	-	this._x)*dt/this._interp) ;
							this._y = this._y + floor((followy	-	this._y)*dt/this._interp) ;


							this._x = (this._x > this._maxx) ? this._maxx : this._x ;
							this._x = (this._x < this._minx) ? this._minx : this._x ;

							this._y = (this._y > this._maxy) ? this._maxy : this._y ;
							this._y = (this._y < this._miny) ? this._miny : this._y ;

			
							this._leftx		=	Math.floor(this._x - this._centrex) ;
							this._topy		=	Math.floor(this._y - this._centrey) ;

						}
					},
	IsAlive:		function()			// this function here allows us to fire Homing Missiles at Camera!
					{
						return true ; 	
					},
	IsFollowing:	function(who)
					{
						return this._follow == who ;
						
					},
	GetTopLeft: 	function()
					{
						return {x:this._leftx, y:this._topy} ;	
					}, 
	GetPosition: 	function()
					{
						return {_x:this._x, _y:this._y, _z:this._z} ;
					},

	X: 				function()
					{
						return this._x ;
					},
	Y: 				function()
					{
						return this._y ;
					},
	Z: 				function()
					{
						return this._z ;
					},
	GetRegion:		function()
					{
						return this.regionHelper.GetRegion(this._x, this._y) ;
					},

	Render:			function(rHelper)
					{
						
					},

	RenderDebug:	function(rHelper)
					{
						var region 	= 	this.GetRegion() ;
						var rcnt 	=	EnemyManager.GetRegionCount() ;
						
						rHelper.drawText("region:"+region+"                                   regionCount "+rcnt+" pos: "+Math.floor(this._x)+","+Math.floor(this._y),this._x+10,this._y,[255,255,255],{vertical:"bottom",horizontal:"left"}) ;
						rHelper.drawCircle(this._x, this._y,10,[255,255,0]) ;
					},
	GetVectorPosition: 	function()
						{
							this._vpos.SetXyzw(this._x,this._y,this._z,0)
							return this._vpos
						},
	
	
}

File: lua/Camera.lua
-- Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt
if (not Camera) then
Camera				=	{}
Camera.className	=	"Camera"
Camera.INTERPRATE	=	0.4
Camera.debug		=	false

	function Camera.Create(screenDim,mapDim,follow,interp)
	
		local camera = {
				follow		=	follow,
				x 			=	0,
				y			=	0,
				z			=	0,
				leftx		=	0,
				topy		=	0,
				time		=	0,
				mapwidth	=	mapDim:X(),
				mapheight	=	mapDim:Y(),
				centrex		=	screenDim:X()/2,
				centrey		=	screenDim:Y()/2,
				interp		=	interp or Camera.INTERPRATE,
				className	=	Camera.className,
		}

		setmetatable(camera, { __index = Camera }) 

		Camera.Init(camera)
		return camera 

	end

	function Camera.SetRestriction(camera,minx,maxx,miny,maxy)
		camera.minx	=	minx
		camera.maxx	=	maxx
		camera.miny	=	miny
		camera.maxy	=	maxy - 64			-- Fudge! Somewrong caclulation
	end
	
	function Camera.Init(camera)
		if (camera.follow) then
			camera.x,camera.y,camera.z 	=	camera.follow:GetPosition()
		end
		
		Camera.SetRestriction(camera,camera.centrex,camera.mapwidth-camera.centrex,camera.centrey,camera.mapheight-camera.centrey)

		camera.vpos		=	Vector4.Create(camera.x,camera.y)

	end

	function Camera.Follow(camera,follow,snap)
		camera.follow		=	follow
		if (snap) then
			local x,y,z	=	follow:GetPosition()
			camera.x	=	x
			camera.y	=	y
		end
		
	end
	
	function Camera.Update(camera,dt)
		camera.time = camera.time + dt  
		if (camera.follow) then
			local followx,followy,followz = camera.follow:GetPosition()
			camera.x = camera.x + math.modf((followx	-	camera.x)*dt/camera.interp)
			camera.y = camera.y + math.modf((followy	-	camera.y)*dt/camera.interp)
--			camera.x	=	camera.follow.x
--			camera.y	=	camera.follow.y

			camera.x = (camera.x > camera.maxx) and camera.maxx or camera.x
			camera.x = (camera.x < camera.minx) and camera.minx or camera.x

			camera.y = (camera.y > camera.maxy) and camera.maxy or camera.y
			camera.y = (camera.y < camera.miny) and camera.miny or camera.y

			
			camera.leftx	=	camera.x - camera.centrex
			camera.rightx	=	camera.x + camera.centrex
			camera.topy		=	camera.y - camera.centrey
			camera.bottomy 	=	camera.y + camera.centrey

			camera.diffx	=	camera.leftx
			camera.diffy	=	camera.topy
			

		end
		
	end
	
	function Camera.GetPosition(camera)
		return camera.x,camera.y,camera.z
	end
	
	function Camera.GetVectorPosition(camera)
		camera.vpos:SetXyzw(camera.x,camera.y,camera.z,0)
		return camera.vpos
	end
	
	-- Used to Render Game Objects - must support the function Render(renderHelper,object,x,y)
	-- which are relative to the camera.
	
	function Camera.RenderObjectFromCameraView(camera,object,renderHelper)
		local x,y,z = object:GetPosition()
		object:Render(renderHelper,x-math.floor(camera.diffx),y - math.floor(camera.diffy))
--		RenderHelper.DrawText(renderHelper,string.format(" TOP CAM POS %f,%f",camera.leftx,camera.topy),600,600)
	end
	
	function Camera.Render(camera,renderHelper)
		if (Camera.debug) then
			RenderHelper.DrawCircle(renderHelper,camera.centrex,camera.centrey,8,{red=127,green=127,blue=40,alpha=0})
			RenderHelper.DrawText(renderHelper,string.format(" CAM POS %f,%f MAX X %f MAX Y %f",camera.x,camera.y,camera.maxx,camera.maxy),camera.centrex,camera.centrey)
		end
	end
	
end
